test/Microsoft.Azure.SignalR.Tests/ServiceConnectionContainerBaseTests.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Azure.SignalR.Protocol;
+using Xunit;
+
+namespace Microsoft.Azure.SignalR.Tests
+{
+    public class ServiceConnectionContainerBaseTests
+    {
+        [Fact]
+        public async Task testIfConnectionWillRestartAfterShutdown()
+        {
+            List<IServiceConnection> connections = new List<IServiceConnection>
+            {
+                new SimpleTestServiceConnection(),
+                new SimpleTestServiceConnection() // A connection which is not in Connected status could be replaced.
+            };
+
+            IServiceConnection connection = connections[1];
+
+            using TestServiceConnectionContainer container = new TestServiceConnectionContainer(connections, factory: new SimpleTestServiceConnectionFactory());
+            container.ShutdownForTest();
+
+            await container.OnConnectionCompleteForTestShutdown(connection);
+
+            // connection should be replaced, but it's StartAsync method should not be called.
+            Assert.NotEqual(container.Connections[1], connection);
+            Assert.NotEqual(ServiceConnectionStatus.Connected, container.Connections[1].Status);
+        }
+
+        private sealed class SimpleTestServiceConnectionFactory : IServiceConnectionFactory
+        {
+            public IServiceConnection Create(HubServiceEndpoint endpoint, IServiceMessageHandler serviceMessageHandler, ServerConnectionType type)
+            {
+                return new SimpleTestServiceConnection();
+            }
+        }
+
+        private sealed class SimpleTestServiceConnection : IServiceConnection
+        {
+            public ServiceConnectionStatus Status { get; set; }
+
+            public Task ConnectionInitializedTask => Task.Delay(TimeSpan.FromSeconds(1));

```

Comment:
Delay 1 sec is too long, maybe 10 milliseconds?

Reply:
Actually we don't have to wait in the current implementation since our `StartCoreAsync` doesn't truly await anything before calling inner `StartAsync`. 

But what if it does something that needs to be awaited, and, won't cost too long in future implementation? (like write a log to file, cause real await but very short)

That's why I introduced a 1s delay in this test case. **1s for 1 test case is fair**, to strike a balance between robust and time-costs.